lib/Sema/CSSimplify.cpp

```diff
@@ -960,9 +960,32 @@ ConstraintSystem::TypeMatchResult constraints::matchCallArguments(
 
   for (unsigned paramIdx = 0, numParams = parameterBindings.size();
        paramIdx != numParams; ++paramIdx){
-    // Skip unfulfilled parameters. There's nothing to do for them.
-    if (parameterBindings[paramIdx].empty())
+    // If a parameter is unfulfilled, validate the default argument if it is a
+    // magic literal expression.
+    if (parameterBindings[paramIdx].empty()) {
+      if (!isa<FuncDecl>(callee))
+        continue;
+
+      auto allParams = dyn_cast<FuncDecl>(callee)->getParameters()->getArray();
+      auto param = allParams[paramIdx];
+
+      assert(param->isDefaultArgument());
+
+      auto defaultValueExpr = param->getDefaultValue();
+
+      if (!isa<MagicIdentifierLiteralExpr>(defaultValueExpr))
+        continue;
+
+      cs.generateConstraints(defaultValueExpr, param->getDeclContext());
+
+      auto defaultTy = cs.getType(defaultValueExpr);
+      auto paramTy = param->getType();
+      cs.addConstraint(
+          ConstraintKind::ArgumentConversion, defaultTy, paramTy,
+          locator.withPathElement(ConstraintLocator::FunctionArgument));
```

Comment:
`FunctionArgument` is great but that's not really what we want here since it'd be great to get a tailored diagnostic for this case and not just an argument conversion one...